neuralpredictors/layers/cores/conv3d.py

```diff
@@ -0,0 +1,503 @@
+            )
+
+            if batch_norm:
+                if independent_bn_bias:
+                    layer["norm"] = nn.BatchNorm3d(hidden_channels, momentum=momentum)
+                else:
+                    layer["norm"] = nn.BatchNorm3d(hidden_channels, momentum=momentum, affine=bias and batch_norm_scale)
+                    if bias:
+                        if not batch_norm_scale:
+                            layer["bias"] = Bias3DLayer(hidden_channels)
+                    elif batch_norm_scale:
+                        layer["scale"] = Scale3DLayer(hidden_channels)
+
+            if self.final_nonlinearity or l < self.layers:
+                if hidden_nonlinearities == "adaptive_elu":
+                    layer["nonlin"] = self.nonlinearities[hidden_nonlinearities](x_shift=x_shift, y_shift=y_shift)
+                else:
+                    layer["nonlin"] = self.nonlinearities[hidden_nonlinearities]()
+
+            self.features.add_module("layer{}".format(l + 1), nn.Sequential(layer))
+
+        self.initialize(cuda=cuda)
+
+    def forward(self, x):
+        for l, features in enumerate(self.features):
+            x = features(x)
+        return x
+
+    def laplace_spatial(self):
+        laplace = 0
+        for filter_index in range(self.features[0].conv.weight.shape[2]):
+            laplace += self._input_weight_regularizer(self.features[0].conv.weight[:, :, filter_index, :, :])
+        return laplace
+
+    def laplace_temporal(self):
+        laplace = 0
+        for w in range(self.features[0].conv.weight.shape[-2]):
+            for h in range(self.features[0].conv.weight.shape[-1]):
+                laplace += self.temporal_regulazirer(self.features[0].conv.weight[:, :, :, w, h])
+        return laplace
+
+    def regularizer(self):
+        return self.gamma_input_spatial * self.laplace_spatial(), self.gamma_input_temporal * self.laplace_temporal()
+
+    @property
+    def out_channels(self):
+        return self.hidden_channels[-1]
+
+    def get_output_shape(self, input_shape, layer=None):
+        ch, t, h, w = input_shape[:]
+
+        if layer is None:
+            layers = self.layers
+        else:
+            layers = layer
+
+        for i in range(layers):
+            if i == 0:
+                kernel_size = self.input_kernel
+                spatial_dilation = self.spatial_dilation
+                temporal_dilation = self.temporal_dilation
+            else:
+                kernel_size = self.hidden_kernel[i - 1]
+                spatial_dilation = 1
+                temporal_dilation = 1
+
+            t = math.floor(t - (kernel_size[0] - 1) * temporal_dilation)
+            h = math.floor((h - (kernel_size[1] - 1) * spatial_dilation - 1) / self.stride[i] + 1)
+            w = math.floor((w - (kernel_size[2] - 1) * spatial_dilation - 1) / self.stride[i] + 1)
+
+        output_shape = self.hidden_channels[layers - 1], t, h, w
+        print("core output shape:", output_shape)
```

Comment:
Is this left over from debugging? Generally using `print` in library code is not recommended. That is what loggers are for.